File: Basic Java Algorithms/BinarySearch.java
import java.io.*;
import java.util.*;

public class BinarySearch {

    static int binarySearch(int[] a, int x) {
        int left = 0, right = a.length - 1;
        while (left <= right) {
            int mid = left + (right - left) / 2;
            if (x == a[mid]) {
                return mid;
            } else if (x < a[mid]) {
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }
        return -1;
    }

    static int linearSearch(int[] a, int x) {
        for (int i = 0; i < a.length; i++) {
            if (a[i] == x)
                return i;
        }
        return -1;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int n = scanner.nextInt();
        int[] a = new int[n];
        for (int i = 0; i < n; i++) {
            a[i] = scanner.nextInt();
        }
        int m = scanner.nextInt();
        int[] b = new int[m];
        for (int i = 0; i < m; i++) {
            b[i] = scanner.nextInt();
        }
        for (int i = 0; i < m; i++) {
            
            System.out.print(binarySearch(a, b[i]) + " ");
        }
    }
}

File: csharp/SearchAlgorithms/BinarySearch.cs
using System;

namespace BinarySearch
{
    class Program
    {

        public static int BinSearch(int[] x, int searchValue)
        {
            // Returns index of searchValue in sorted array x, or -1 if not found
            int left = 0;
            int right = x.Length;
            return binarySearch(x, searchValue, left, right);
        }

        public static int binarySearch(int[] x, int searchValue, int left, int right)
        {
            if (right < left)
            {
                return -1;
            }
            int mid = (left + right) >> 1;
            if (searchValue > x[mid])
            {
                return binarySearch(x, searchValue, mid + 1, right);
            }
            else if (searchValue < x[mid])
            {
                return binarySearch(x, searchValue, left, mid - 1);
            }
            else
            {
                return mid;
            }
        }

        static void Main()
        {
            //first you need to sort your array
            int[] A = { 1, 2, 4, 6, 7, 9, 11, 42, 300, 302 }; // Random test data
            int result = BinSearch(A, 9);
            if (result==-1)
            {
                Console.WriteLine("Value not exists");
            }
            else
            {
                string resultString = "Value on index " + result;
                Console.WriteLine(resultString);
            }

        }
    }
}
